### PR TITLE
Add automatic parameter tuning based on minimum cross-entropy thresholding

### PR DESCRIPTION
**Pull request by joeljonsson**
Adds a new method to automatically compute thresholds for the gradient and local intensity scale using Li's iterative Minimum Cross Entropy algorithm ([https://doi.org/10.1016/S0167-8655(98)00057-9](https://doi.org/10.1016/S0167-8655(98)00057-9)). 

The thresholding method is applied only to gradient / intensity scale values where the image intensity exceeds the intensity threshold. This allows handling some difficult cases, e.g. where there are significant regions of background with and without autofluorescence, by manually setting the intensity threshold to an appropriate value.

I have tested this on a number of datasets, and it seems to mostly work very well. Would appreciate some further testing and feedback on this.

### CODE REVIEW

**Comment by cheesema** [comment]
Where did the inspiration come from?

**Comment by joeljonsson** [comment]
> Where did the inspiration come from?

Figured this was long overdue. I just tested any (simple) histogram thresholding method I could find via Python and this one consistently gave good results

File: src/algorithm/APRConverter.hpp [hunk 1]
```diff
@@ -676,6 +683,155 @@ void APRConverter<ImageType>::autoParameters(const PixelData<T> &localIntensityS
...
+#endif
+        for(size_t idx = 0; idx < num_elements; ++idx) {
+            grad_subsampled[idx] = grad_foreground[idx*delta];
+            lis_subsampled[idx] = lis_foreground[idx*delta];
+        }
+    }
+    fine_grained_timer.stop_timer();
+
+    fine_grained_timer.start_timer("autoparameters: compute gradient threshold");
+    par.grad_th = threshold_li(grad_subsampled);
```
Comment (cheesema) [inline src/algorithm/APRConverter.hpp:819]: This par object is taken directly from the class, then we have also one where we are adding to the APR? Maybe we should remove the one from the class itself, and have this function work in a more static way so its more useful in the future? Would require changing of the signiture.

File: src/algorithm/APRConverter.hpp [hunk 2]
```diff
@@ -676,6 +683,155 @@ void APRConverter<ImageType>::autoParameters(const PixelData<T> &localIntensityS
...
+    fine_grained_timer.start_timer("autoparameters: compute sigma threshold");
+    par.sigma_th = threshold_li(lis_subsampled);
+    fine_grained_timer.stop_timer();
+
+    std::cout << "Used parameters: " << std::endl;
+    std::cout << "I_th: " << par.Ip_th << std::endl;
+    std::cout << "sigma_th: " << par.sigma_th << std::endl;
+    std::cout << "grad_th: " << par.grad_th << std::endl;
+    std::cout << "relative error (E): " << par.rel_error << std::endl;
+    std::cout << "lambda: " << par.lambda << std::endl;
```
Comment (cheesema) [inline src/algorithm/APRConverter.hpp:831]: Should we remove this, or make this linked to a verbose flag of some kind?

File: src/algorithm/APRConverter.hpp [hunk 3]
```diff
@@ -676,6 +683,155 @@ void APRConverter<ImageType>::autoParameters(const PixelData<T> &localIntensityS
...
+                lis_foreground[counter] = localIntensityScale.mesh[idx];
+                counter++;
+            }
+        }
+
+        grad_foreground.resize(counter);
+        lis_foreground.resize(counter);
+
+        /// Then we uniformly subsample these signals, as we typically don't need all elements to compute the thresholds
+        const size_t num_elements = std::min((size_t)32*512*512, counter);
```
Comment (cheesema) [inline src/algorithm/APRConverter.hpp:803]: Where do these magic numbers come from?

File: src/algorithm/APRConverter.hpp [hunk 4]
```diff
@@ -459,7 +462,11 @@ inline bool APRConverter<ImageType>::get_apr(APR &aAPR, PixelData<T>& input_imag
     computation_timer.start_timer("apply_parameters");
 
     if( par.auto_parameters ) {
```
Comment (cheesema) [inline src/algorithm/APRConverter.hpp:464]: Would we then put this as default to true?

**Review by cheesema — CHANGES_REQUESTED** [review]
Added some comments, does this work well in all dimensions? have you done any tests on 1D + 2D?

And sorry for taking so long :(

File: src/algorithm/APRConverter.hpp (hunk 1 above)
Comment (joeljonsson) [reply src/algorithm/APRConverter.hpp:819]: All converter methods (should) use the parameters instance of the converter class. The instance in the APR class is mainly there to save the used parameters. But I agree it's slightly confusing...

I can make the function take a parameters object as input, but removing the parameters instance from the converter class would require some refactoring. This would also induce breaking changes, as instead of setting the converter parameters before conversion one would have to either a) set the parameters of the APR object, or b) input a parameters object to all converter methods.

File: src/algorithm/APRConverter.hpp (hunk 2 above)
Comment (joeljonsson) [reply src/algorithm/APRConverter.hpp]: yup! this is done now

File: src/algorithm/APRConverter.hpp (hunk 3 above)
Comment (joeljonsson) [reply src/algorithm/APRConverter.hpp:803]: Completely arbitrary. Perhaps it would make more sense to use a fraction of the image size?

File: src/algorithm/APRConverter.hpp (hunk 4 above)
Comment (joeljonsson) [reply src/algorithm/APRConverter.hpp:464]: I'm not sure, as this would require explicitly turning it off every time you want to set your own values. Perhaps we can automatically turn it on based on some default values (e.g. -1) of `sigma_th` and `grad_th`? This could also allow some more flexibility in terms of which parameters are computed

File: src/algorithm/APRConverter.hpp (hunk 1 above)
Comment (cheesema) [reply src/algorithm/APRConverter.hpp:819]: Fair enough agreed.

File: src/algorithm/APRConverter.hpp (hunk 3 above)
Comment (cheesema) [reply src/algorithm/APRConverter.hpp]: I think this is fine for now, you use the random subsampling.

**Comment by cheesema** [comment]
Everything is good from my side, I added a few minor changes for you to look at, after that feel free to merge.

**Review by cheesema — APPROVED** [review]
